kernel/sched.c

```diff
@@ -168,15 +187,20 @@ static void update_cache(int preempt_ok)
 #ifndef CONFIG_SMP
 	struct k_thread *th = next_up();
 
-	if (_current && !_is_idle(_current) && !_is_thread_dummy(_current)) {
-		/* Don't preempt cooperative threads unless the caller allows
-		 * it (i.e. k_yield())
-		 */
-		if (!preempt_ok && !_is_preempt(_current) && !is_metairq(th)) {
-			th = _current;
-		}
+	if (should_preempt(th, preempt_ok)) {
```

Comment:
If the previous `should_preempt()` call is a no-op, then it can return `th` or `_current` instead of `0` or `1` (after a rename, of course), shaving off a branch here: `_kernel.ready_q.cache = should_preempt(th, preempt_ok);`